src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

```diff
@@ -118,12 +114,45 @@ class ResourceService(private val resourceTypes: Map[ResourceTypeName, ResourceT
     }
   }
 
-  private def validateActionsAndRoles(resourceType: ResourceType, policyMembership: AccessPolicyMembership) = {
-    val invalidAction = policyMembership.actions.find(a => !resourceType.actionPatterns.exists(_.matches(a)))
-    if (invalidAction.isDefined)
-      throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"You have specified an invalid action for resource type ${resourceType.name}. Valid actions are: ${resourceType.actionPatterns.mkString(", ")}"))
-    if (!policyMembership.roles.subsetOf(resourceType.roles.map(_.roleName)))
-      throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"You have specified an invalid role for resource type ${resourceType.name}. Valid roles are: ${resourceType.roles.map(_.roleName).mkString(", ")}"))
+  private def mapEmailsToSubjects(workbenchEmails: Set[WorkbenchEmail]): Future[Map[WorkbenchEmail, Option[WorkbenchSubject]]] = {
+    val eventualSubjects = workbenchEmails.map { workbenchEmail =>
+      directoryDAO.loadSubjectFromEmail(workbenchEmail).map(workbenchEmail -> _)
+    }
+
+    Future.sequence(eventualSubjects).map(_.toMap)
+  }
+
+  // When validating the policy, we want to collect each entity that was problematic and report that back using ErrorReports
+  private def validatePolicy(resourceType: ResourceType, policyMembership: AccessPolicyMembership, members: Map[WorkbenchEmail, Option[WorkbenchSubject]]) = {
+    val validationErrors = validateMemberEmails(members) ++ validateActions(resourceType, policyMembership) ++ validateRoles(resourceType, policyMembership)
+    if (validationErrors.nonEmpty)
+      throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "You have specified an invalid policy", validationErrors.toSeq))
+  }
+
+  // Keys are the member email addresses we want to add to the policy, values are the corresponding result of trying to lookup the
+  // subject in the Directory using that email address.  If we failed to find a matching subject, then that's not good
+  private def validateMemberEmails(emailsToSubjects: Map[WorkbenchEmail, Option[WorkbenchSubject]]): Option[ErrorReport] = {
+    val invalidEmails = for(tuple <- emailsToSubjects if tuple._2.isEmpty) yield tuple._1
```

Comment:
but better yet look at emailsToSubjects.collect

Reply:
@dvoet after playing with this quite a bit, I'm not sure `.collect` is going to give us what we want as easily as the `for` comprehension does:

```
@ m 
res22: Map[Int, Option[String]] = Map(1 -> Some("A"), 2 -> None, 3 -> None, 4 -> Some("D"))

@ m.collect {case (k,v) => if(v.isEmpty) k->None} 
res23: collection.immutable.Iterable[Any] = List((), (2, None), (3, None), ())
```
Note in the above example that I was able to filter out the right tuples, but I still have additional filtering to do to whittle that List down to the `List(2, 3)` that I was after.  Whereas the `for` comprehension gets me there pretty easily:
```
@ for((k,v) <- m if v.isEmpty) yield k 
res28: collection.immutable.Iterable[Int] = List(2, 3)
```

This may be just because I'm new to Scala, but I think that a `for` comprehension is clearer here because it does the `map` and `filter` to give me back just the offending emails I'm after.

For now, I've pushed a change using your suggested improvement to the `for` comprehension.

Happy to look at this again with you if there's something I've missed.

Reply:
Nevermind, @breilly2 showed me how to use the partial function properly for what we're trying to do here and as a result the `.collect` statement works.